src/main/java/com/es/phoneshop/web/ProductDetailPageServlet.java

```diff
@@ -2,31 +2,81 @@
 
 import com.es.phoneshop.dao.ProductDao;
 import com.es.phoneshop.dao.impl.ArrayListProductDao;
+import com.es.phoneshop.exceptions.OutOfStockException;
+import com.es.phoneshop.model.cart.Cart;
+import com.es.phoneshop.model.cart.CartService;
+import com.es.phoneshop.model.cart.DefaultCartService;
+import com.es.phoneshop.model.recentlyViewedProducts.DefaultRecentlyViewedProductsService;
+import com.es.phoneshop.model.recentlyViewedProducts.RecentlyViewedProducts;
+import com.es.phoneshop.model.recentlyViewedProducts.RecentlyViewedProductsService;
 
 import javax.servlet.ServletConfig;
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
 import java.util.NoSuchElementException;
 
 public class ProductDetailPageServlet extends HttpServlet {
     private ProductDao productDao;
+    private CartService cartService;
+    private RecentlyViewedProductsService recentlyViewedProductsService;
 
     @Override
     public void init(ServletConfig config) throws ServletException {
         productDao = ArrayListProductDao.getInstance();
+        cartService = DefaultCartService.getInstance();
+        recentlyViewedProductsService = DefaultRecentlyViewedProductsService.getInstance();
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
         try {
+            RecentlyViewedProducts recentlyViewedProducts =
+                    recentlyViewedProductsService.getRecentlyViewedProducts(request);
+            request.setAttribute("recentProducts", recentlyViewedProducts.getRecentlyViewedProducts());
             Long productId = Long.valueOf(request.getPathInfo().substring(1));
             request.setAttribute("product", productDao.getProduct(productId));
+            Cart cart = cartService.getCart(request);
+            request.setAttribute("cart", cart);
             request.getRequestDispatcher("/WEB-INF/pages/productDetails.jsp").forward(request, response);
+            recentlyViewedProductsService.addProduct(productId, recentlyViewedProducts);
```

Comment:
Add product **before** setting recentProducts attribute to request (so that the product being currently visited also shows up in the list)